pkg/ddl/syncer/syncer.go

```diff
@@ -126,16 +128,98 @@ type SchemaSyncer interface {
 	// the latest schema version. (exclude the isolated TiDB)
 	// It returns until all servers' versions are equal to the latest version.
 	OwnerCheckAllVersions(ctx context.Context, jobID int64, latestVer int64) error
+	// SyncJobSchemaVerLoop syncs the schema versions on all TiDB nodes for DDL jobs.
+	SyncJobSchemaVerLoop(ctx context.Context)
 	// Close ends SchemaSyncer.
 	Close()
 }
 
+// nodeVersions is used to record the schema versions of all TiDB nodes for a DDL job.
+type nodeVersions struct {
+	sync.Mutex
+	nodeVersions map[string]int64
+	// onceMatchFn is used to check if all the servers report the least version.
+	// If all the servers report the least version, i.e. return true, it will be
+	// set to nil.
+	onceMatchFn func(map[string]int64) bool
+}
+
+func newNodeVersions(initialCap int, fn func(map[string]int64) bool) *nodeVersions {
+	return &nodeVersions{
+		nodeVersions: make(map[string]int64, initialCap),
+		onceMatchFn:  fn,
+	}
+}
+
+func (v *nodeVersions) add(nodeID string, ver int64) {
+	v.Lock()
+	defer v.Unlock()
+	v.nodeVersions[nodeID] = ver
+	if v.onceMatchFn != nil {
+		if ok := v.onceMatchFn(v.nodeVersions); ok {
+			v.onceMatchFn = nil
+		}
+	}
+}
+
+func (v *nodeVersions) del(nodeID string) {
+	v.Lock()
+	defer v.Unlock()
+	delete(v.nodeVersions, nodeID)
+	// we don't call onceMatchFn here, for only "add" can cause onceMatchFn return
+	// true currently.
+}
+
+func (v *nodeVersions) len() int {
+	v.Lock()
+	defer v.Unlock()
+	return len(v.nodeVersions)
+}
+
+func (v *nodeVersions) matchOrSet(fn func(nodeVersions map[string]int64) bool) {
+	v.Lock()
+	defer v.Unlock()
+	if ok := fn(v.nodeVersions); !ok {
+		// onceMatchFn must be nil before.
```

Comment:
maybe move it to function comments, so when I hover the cursor above the function in IDE I can see the comments.